src/ui/components/contacts/ContactsPage.jsx

```diff
@@ -0,0 +1,149 @@
+import React, {useState, useEffect} from 'react';
+import PropTypes, {shape, string} from 'prop-types';
+import useSWR from 'swr';
+import PageWrapper from '../common/PageWrapper';
+import NavigationSidebar from '../common/navigation/NavigationSidebar';
+import {MdPersonAddAlt1, MdCheckCircle} from 'react-icons/md';
+import PrimaryButton from '../common/buttons/PrimaryButton';
+import {
+  Flex,
+  Heading,
+  Icon,
+  Text,
+  useBreakpointValue,
+  useDisclosure,
+  useToast,
+} from '@chakra-ui/react';
+import ContactsTable from './ContactsTable';
+import ContactsModal from './ContactsModal';
+import LoadingSpinner from '../common/LoadingSpinner';
+import PrimaryToast from '../common/PrimaryToast';
+
+const fetcher = (url, token) =>
+  fetch(url, {
+    headers: {Authorization: `Bearer ${token}`, 'Content-Type': 'application/json'},
+  }).then(res => res.json());
+
+export default function ContactsPage({session}) {
+  const {jwt} = session;
+  const {data: contactsData, error: contactsError} = useSWR(['/api/contacts', jwt], fetcher);
+  const {data: companiesData, error: companiesError} = useSWR(['/api/companies', jwt], fetcher);
+  const [contacts, setContacts] = useState([]);
+  const [companies, setCompanies] = useState([]);
+  const [modalType, setModalType] = useState('New');
+  const [currentContact, setCurrentContact] = useState(null);
+
+  const showSidebar = useBreakpointValue({
+    base: false,
+    lg: true,
+  });
+  const {isOpen, onOpen, onClose} = useDisclosure();
+
+  const toast = useToast();
+  const renderToast = (title, description) =>
+    toast({
+      position: 'top',
+      duration: 3000,
+      render: () => <PrimaryToast title={title} description={description} icon={MdCheckCircle} />,
+    });
+
+  const addContact = contact => {
+    setContacts([...contacts, contact]);
+    renderToast('Contact Added Successfully.', `${contact.name} has been added to your contacts.`);
+  };
+
+  const editContact = contact => {
+    const updatedContacts = contacts.map(con => {
+      if (con.id === contact.id) con = contact;
+      return con;
```

Comment:
Could perhaps use filter here. Not a blocker.

Reply:
This function is called when a user updates one of their contacts. I'm taking the existing list of contacts and replacing the old contact with its updated version so its place in the table remains the same and the data is updated. I don't think I can accomplish that with filter because the list of contacts is part of the component state and I need to pass a list of all contacts to update the state.